### PR TITLE
[Nullability Annotations to Java Classes] Use Updated and Null Proof `MediaModel` Class (`breaking`)

### PR DESCRIPTION
**Pull request by ParaskP7**
Parent [FluxC#2799](https://github.com/wordpress-mobile/WordPress-FluxC-Android/issues/2799)
Accompanying [FluxC#2886](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2886)

-----

### Description

This PR accompanying the above FluxC PR, and uses an updated and null-proof [MediaModel.java](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/7fd7840a9bd900fde8e32ad8bd5632ad8f05dff3/fluxc/src/main/java/org/wordpress/android/fluxc/model/MediaModel.java#L4) class.

FYI: This change is `breaking`, meaning that any client that depended on the [MediaModel.java](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/7fd7840a9bd900fde8e32ad8bd5632ad8f05dff3/fluxc/src/main/java/org/wordpress/android/fluxc/model/MediaModel.java#L4) model should update to the new APIs (`non-default constructors`) as the existing API (`default constructor`) are now deprecated. As such, this change is inherently `risky`, meaning that there are compile-time changes associated with this change and thus needs testing to verify correctness, both on the library's and client's side.

-----

@0nko I added you as the main reviewer, not so randomly (it being a continuation of [FluxC#2878](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2878)), since I just wanted someone from the WooCommerce mobile team to be aware of and sign-off on that change. Feel free to merge this PR directly yourself if you deem so.

-----

FluxC Update List:

1. [Update fluxc version to pr hash (2886)](https://github.com/woocommerce/woocommerce-android/pull/10078/commits/faf148bae26637eff20f4732f792b277472d1170)
2. [Update fluxc version to pr hash (2886)](https://github.com/woocommerce/woocommerce-android/pull/10078/commits/430a47d36364d4d2ffbc62ca71a2f5ecc0d4f1aa)
3. [Update fluxc version to trunk hash (2886)](https://github.com/woocommerce/woocommerce-android/pull/10078/commits/2a44866fd36f8504df6c0a4fa72625c0705c838a)
4. [Update fluxc version to trunk hash (2886)](https://github.com/woocommerce/woocommerce-android/pull/10078/commits/66ee06e007af315baf2658f36c45877f6fba2d09)

Deprecation Resolution List (`breaking`):

1. [Resolve media model constructor deprecation warnings](https://github.com/woocommerce/woocommerce-android/pull/10078/commits/a06e525683e3efa15200703f63f3101637c5c1dd)
5. [Make media model parameter nullable on media upload exception](https://github.com/woocommerce/woocommerce-android/pull/10078/commits/afa7aa793be8228e8cf927b1d6d5fcfd3f0d37ba)

Nullability Checks List:

1. [Un-guard usages of non-null media model url](https://github.com/woocommerce/woocommerce-android/pull/10078/commits/508ffcd85933625cd1306fffad86e3d1f0d6956f)
6. [Guard usages of nullable media model file path](https://github.com/woocommerce/woocommerce-android/pull/10078/commits/73a45bab6ad64f943000f59156e497d0340c8f5e)

-----

#### (`REST`):

- Smoke test any media related functionality and see if everything is working as expected. For a couple of examples, you can expand and follow the inner and more explicit test steps within:

#### (`XMLRPC`):

- Create a new self-hosted WordPress site for XMLRPC testing purposes ([jurassic.ninja](https://jurassic.ninja/)).
- Smoke test any media related functionality and see if everything is working as expected. For a couple of examples, you can expand and follow the inner and more explicit test steps within:

<details>
    <summary>Photos Screens [ProductDetailFragment.java + ProductImagesFragment.kt + ProductImageViewerFragment.kt]</summary>

- Go to `Products` tab and tap on any product.
- Tap on any photo on top and verify that the `Photos` screen is shown and functioning as expected.
- For example try adding a new photo, select an upload method, and wait for the photo to be uploaded.
- Tap on the photo preview and verify that the `Photo Preview` screen is shown and functioning as expected.
- For example try swipe left-right to navigate between photo previews.

</details>

-----

### Merge instructions:

1. [x] Wait for the accompanying [FluxC#2886](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2886) PR to be merged.
2. [x] Update `fluxCVersion` to point to the `trunk` hash that includes the above solution.
3. [x] Remove the `status: do not merge` label.
7. [x] Merge this PR.

-----

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/media/MediaFilesRepository.kt [hunk 1]
```diff
@@ -144,21 +144,22 @@ class MediaFilesRepository @Inject constructor(
                 }
 
                 event.completed -> {
-                    val channelResult = if (event.media?.url != null) {
-                        WooLog.i(T.MEDIA, "MediaFilesRepository > uploaded media ${event.media?.id}")
+                    val media = event.media
+                    val channelResult = if (media != null && media.url.isNotBlank()) {
```
Comment (ParaskP7) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/media/MediaFilesRepository.kt:148]: Warning (⚠️): I am not fully confident about this change here as I am not sure which check is better, the version I chose here:

```kotlin
if (media != null && media.url.isNotBlank()) { UploadSuccess(media) }
```

Or another version of it, which just checks for a nullable `media`, before marking this as upload success, but without checking for a blank URL:

```kotlin
if (media != null) { UploadSuccess(media) }
```

FYI: Mind the fact `media.url` can no longer be nullable with the updated `media` model.

Let me know your thoughts on this as I fear this has the potential to change the behavior somehow, on cases that we think a `media` is still valid even with a blank URL, thank YOU for helping! 🙏

Comment (0nko) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/media/MediaFilesRepository.kt:148]: I wasn't sure either so I checked for the usage of `MediaModel.url` and it seems like we don't use/check for the empty value anywhere, so the the 1st option should be fine. Besides, uploading an empty URL doesn't make sense anyway :).

**Review by 0nko — APPROVED** [review]
I've checked the code and tested both REST and XML-RPC variants and everything seems to be working fine. However, before we proceed with the merge, we'll need to also update the [MediaPicker](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android) library.

Nice work @ParaskP7! 🥇

File: WooCommerce/src/main/kotlin/com/woocommerce/android/media/MediaFilesRepository.kt (hunk 1 above)
Comment (ParaskP7) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/media/MediaFilesRepository.kt:148]: Awesome, thanks for checking, verifying and clarifying this @0nko , much appreciated! 🙇 🙏 

FYI: I'll then stick with the version we have here then. 👍

**Comment by ParaskP7** [comment]
Thanks you so much for review, testing and approving this and the [FluxC PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2886) @0nko , you rock! 🙇 ❤️ 

> I've checked the code and tested both REST and XML-RPC variants and everything seems to be working fine. However, before we proceed with the merge, we'll need to also update the [MediaPicker](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android) library.

Oh, I wasn't sure about that, thinking that [MediaPicker](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android) will compile/work anyway (for now), just because it will be pointing to a previous version of [FluxC](https://github.com/wordpress-mobile/WordPress-FluxC-Android). And also, that only when we'll need to update [FluxC](https://github.com/wordpress-mobile/WordPress-FluxC-Android) there that we would also need to deal with the `MediaModel` update. 🤔

But if you say so, I'll try to create another PR on the [MediaPicker](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android) side, deal with the `MediaModel` update there too and update the `mediapickerVersion` on this PR to point to that PR hash for testing purposes. It is indeed making sense to deal with that too now, rather than later. 💯 

Would you be willing to go through another round of testing afterwards? 🙏

**Comment by ParaskP7** [comment]
👋 @0nko !

> But if you say so, I'll try to create another PR on the [MediaPicker](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android) side, deal with the MediaModel update there too and update the mediapickerVersion on this PR to point to that PR hash for testing purposes. It is indeed making sense to deal with that too now, rather than later. 💯

A quick update on that! 👀 

I just checked the [MediaPicker](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android) library and noticed that the `wordPressFluxCVersion` version is pointing to `develop-fa819801570505a0e9b4f7f226bb704b88ccc1d2` (😮 `develop` 🤷), a version which is 3 years old ([commit](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android/commit/00947a9086ecdf2192b0bdc95a22e1647696d0b1)). 😨

As such, I think it is better for me to NOT go through that change and let someone else deal with the `wordPressFluxCVersion` version update on [MediaPicker](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android). Update `wordPressFluxCVersion` version now might cause additional breaking changes, unrelated to this `MediaModel` update and I don't know if we should mix those 2 together. Wdyt? 🤔

**Comment by 0nko** [comment]
> Would you be willing to go through another round of testing afterwards? 🙏

@ParaskP7 Absolutely 👍. Thanks for handling it!

**Comment by 0nko** [comment]
> As such, I think it is better for me to NOT go through that change and let someone else deal with the wordPressFluxCVersion version update on [MediaPicker](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android). Update wordPressFluxCVersion version now might cause additional breaking changes, unrelated to this MediaModel update and I don't know if we should mix those 2 together. Wdyt? 🤔

Just saw your comment. OK, let's get these changes merged and I can take a look at the library afterwards.

**Comment by ParaskP7** [comment]
> Just saw your comment. OK, let's get these changes merged and I can take a look at the library afterwards.

Great, thanks for agreeing to that @0nko and I am sorry to delegate that to you, that's wasn't my intention here. 🙏 

Btw, I quickly searched for `MediaModel` and `MediaStore` within [MediaPicker](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android) and didn't notice anything too breaking. This library is never constructing a `MediaModel` class itself, just using an existing one from the `MediaStore` to then construct a `MediaItem` from it (see [here](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android/blob/22d782f51c2b5c2848497cc5ad53c78f8fc9d805/mediapicker/source-wordpress/src/main/java/org/wordpress/android/mediapicker/source/wordpress/MediaLibrarySource.kt#L124-L140)). Thus, I think `MediaModel` will not be too much of a breaking change on this library, or at least anything hard for you to update after first updating [FluxC](https://github.com/wordpress-mobile/WordPress-FluxC-Android). 🤞 

But, I do indeed worry about the [FluxC](https://github.com/wordpress-mobile/WordPress-FluxC-Android) library that is used there, which is 3 year old, and thus don't want to update that, as part of this work, just to deal with the `MediaModel` update. I hope that justifies why I suggest going this way. ☺️

**Comment by ParaskP7** [comment]
FYI: I'll resolve the `build.gradle` merge conflict when I'll be updating `fluxCVersion` to point to the `trunk` hash that includes the above solution (see `Merge instructions`).

**Comment by 0nko** [comment]
@ParaskP7 I tried to update the FluxC reference and get it merged but I can't build the project. I'm getting a useless `e: Could not load module <Error module>` error message. I was wondering if it's because the dependency catalogue is using an old version of Kotlin (1.6). But it seems like you were more successful.

**Comment by ParaskP7** [comment]
👋 @0nko and thanks for trying to find the correct FluxC reference and then do the merge yourself. 🙇 

> I'm getting a useless e: Could not load module <Error module> error message. I was wondering if it's because the dependency catalogue is using an old version of Kotlin (1.6).

FYI: This was most like related to this `CGPNUU63E/p1699440897019819` internal discussion of our and [FluxC#2890](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2890).

Let me know if that makes sense. 🙏 

> But it seems like you were more successful.

🎉  🚀

**Pull request merged**